await.go

```diff
@@ -0,0 +1,201 @@
+	}
+	cp, err := a.await(ctx, i)
+	return i, cp, err
+}
+
+// pollLoop MUST be called in a goroutine when constructing an IntegrationAwaiter
+// and will run continually until its context is cancelled. It wakes up every
+// `pollPeriod` to check if there are clients blocking. If there are, it requests
+// the latest checkpoint from the log, parses the tree size, and releases all clients
+// that were blocked on an index smaller than this tree size.
+func (a *IntegrationAwaiter) pollLoop(ctx context.Context, readCheckpoint func(ctx context.Context) ([]byte, error), pollPeriod time.Duration) {
+	for {
+		select {
+		case <-ctx.Done():
+			klog.Info("IntegrationAwaiter exiting due to context completion")
+			return
+		case <-time.After(pollPeriod):
+			// It's worth this small lock contention to make sure that no unnecessary
+			// work happens in personalities that aren't performing writes.
+			a.mu.Lock()
+			hasClients := a.waiters.Front() != nil
+			a.mu.Unlock()
+			if !hasClients {
+				continue
+			}
+		}
+		// Note that for now, this releases all clients in the event of a single failure.
+		// If this causes problems, this could be changed to attempt retries.
+		rawCp, err := readCheckpoint(ctx)
+		if err != nil {
+			a.releaseClientsErr(fmt.Errorf("readCheckpoint: %v", err))
+			continue
+		}
+		// Parsing a checkpoint like this is only acceptable because we're in the same binary as the
+		// log implementation that generated it and thus we can safely assume it's a well formed and
+		// validly signed checkpoint. Anyone copying similar logic into client code will get hurt.
+		parts := bytes.SplitN(rawCp, []byte{'\n'}, 3)
+		if want, got := 3, len(parts); want != got {
+			a.releaseClientsErr(fmt.Errorf("invalid checkpoint: %q", rawCp))
+			continue
+		}
+		sizeStr := string(bytes.SplitN(rawCp, []byte{'\n'}, 4)[1])
+		size, err := strconv.ParseUint(sizeStr, 10, 64)
+		if err != nil {
+			a.releaseClientsErr(fmt.Errorf("failed to turn checkpoint size of %q into uint64: %v", sizeStr, err))
+			continue
+		}
+		a.releaseClients(size, rawCp)
+	}
+}
+
+func (a *IntegrationAwaiter) await(ctx context.Context, i uint64) ([]byte, error) {
+	a.mu.Lock()
+	if a.size > i {
+		cp := a.checkpoint
+		a.mu.Unlock()
+		return cp, nil
+	}
+	done := make(chan checkpointOrDeath, 1)
+	w := waiter{
+		index:  i,
+		result: done,
+	}
+	a.waiters.PushBack(w)
+	a.mu.Unlock()
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case cod := <-done:
+		return cod.cp, cod.err
+	}
+}
+
+func (a *IntegrationAwaiter) releaseClientsErr(err error) {
+	cod := checkpointOrDeath{
+		err: err,
+	}
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	for e := a.waiters.Front(); e != nil; e = e.Next() {
+		w := e.Value.(waiter)
+		w.result <- cod
+		close(w.result)
+		a.waiters.Remove(e)
+	}
+}
+
+func (a *IntegrationAwaiter) releaseClients(size uint64, cp []byte) {
+	cod := checkpointOrDeath{
+		cp: cp,
+	}
+	a.mu.Lock()
+	a.size = size
+	a.checkpoint = cp
+	defer a.mu.Unlock()
```

Comment:
nit: move this up to immediately after `Lock`?

Reply:
Done